File: backend-rust/src/socket/events.rs
use std::sync::LazyLock;

use dashmap::DashMap;
use serde_json::{json, Value};
use socketioxide::extract::{Data, SocketRef};
use tracing::info;
use crate::state::AppState;

use super::guest_ai;

static ROOM_SESSIONS: LazyLock<DashMap<String, Vec<String>>> = LazyLock::new(DashMap::new);

pub fn register_events(socket: &SocketRef, state: &AppState) {
    socket.on("join", {
        let st = state.clone();
        move |s: SocketRef, Data::<Value>(data)| {
            let st = st.clone();
            async move { handle_join(&st, &s, data).await }
        }
    });

    socket.on("join_room", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(room)| {
            let st = st.clone();
            async move { handle_join_room(&st, &s, &room).await }
        }
    });

    socket.on("chat", {
        let st = state.clone();
        move |s: SocketRef, Data::<Value>(data)| {
            let st = st.clone();
            async move { handle_chat(&st, &s, data).await }
        }
    });

    socket.on("chat_history", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(room)| {
            let st = st.clone();
            async move { handle_chat_history(&st, &s, &room).await }
        }
    });

    socket.on("chat_history_guest", {
        let st = state.clone();
        move |s: SocketRef, Data::<Value>(data)| {
            let st = st.clone();
            async move { handle_chat_history_guest(&st, &s, data).await }
        }
    });

    socket.on("chat_history_app", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move { handle_chat_history_app(&st, &s, &app_id).await }
        }
    });

    socket.on("chat_mark_read", move |s: SocketRef, Data::<Value>(data)| async move {
        let _ = data;
        let _ = s.emit("chat_mark_read", &json!({"ok": true}));
    });

    socket.on("chat_mark_all_read", move |s: SocketRef, Data::<String>(room)| async move {
        let _ = room;
        let _ = s.emit("chat_mark_all_read", &json!({"ok": true}));
    });

    socket.on("chat_recall_message", move |s: SocketRef, Data::<Value>(data)| async move {
        let _ = s.broadcast().emit("chat_message_recalled", &data);
    });

    socket.on("user_typing", move |s: SocketRef, Data::<Value>(data)| async move {
        let _ = s.broadcast().emit("user_typing", &data);
    });

    socket.on("broadcast_notification", move |s: SocketRef, Data::<Value>(data)| async move {
        let _ = s.broadcast().emit("notification", &data);
    });

    socket.on("register_guest_phone", {
        let st = state.clone();
        move |s: SocketRef, Data::<Value>(data)| {
            let st = st.clone();
            async move { guest_ai::process_register_guest_phone(&st, &s, data).await }
        }
    });

    socket.on("chat_list_users", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move {
                let _ = s.emit("chat_list_users", &list_users(&st, &app_id));
            }
        }
    });

    socket.on("chat_guests_list", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move {
                let _ = s.emit("chat_guests_list", &guest_ai::list_guest_sessions(&st, &app_id));
            }
        }
    });

    socket.on("chat_list_rooms", move |s: SocketRef, Data::<String>(app_id)| async move {
        let rooms: Vec<String> = ROOM_SESSIONS.iter().map(|e| e.key().clone()).collect();
        let _ = s.emit("chat_list_rooms", &json!({"appId": app_id, "rooms": rooms}));
    });

    socket.on("chat_create_room", move |s: SocketRef, Data::<String>(payload)| async move {
        let _ = s.emit("chat_create_room", &json!({"room": payload, "created": true}));
    });

    socket.on("csm_sign_in", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(encrypted)| {
            let st = st.clone();
            async move { handle_csm_sign_in(&st, &s, &encrypted).await }
        }
    });

    socket.on("csm_msg_update", move |s: SocketRef, Data::<Value>(data)| async move {
        let _ = s.broadcast().emit("csm_msg_update", &data);
    });

    socket.on("csm_register_an_account", move |s: SocketRef, Data::<String>(encrypted)| async move {
        let _ = encrypted;
        let _ = s.emit("csm_register_an_account", &json!({"ok": true}));
    });

    socket.on("request_chat_history_app_snapshot", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move { handle_chat_history_app(&st, &s, &app_id).await }
        }
    });

    socket.on("chat_list_online_admins", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move {
                let _ = s.emit("chat_list_online_admins", &list_users(&st, &app_id));
            }
        }
    });

    socket.on("chat_list_app_users_presence", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move {
                let _ = s.emit("chat_list_app_users_presence", &json!({"appId": app_id, "online": []}));
            }
        }
    });

    socket.on("chat_list_groups", {
        let st = state.clone();
        move |s: SocketRef, Data::<String>(app_id)| {
            let st = st.clone();
            async move {
                let page = st.record_manager.filter(&app_id, "csm_group_members", &Default::default());
                let _ = s.emit("chat_list_groups", &page.get("data").cloned().unwrap_or(json!([])));
            }
        }
    });

    socket.on_disconnect(|s: SocketRef| async move {
        guest_ai::on_disconnect(&s.id.to_string());
        info!("Socket disconnected: {}", s.id);
    });
}

async fn handle_join(state: &AppState, socket: &SocketRef, data: Value) {
    guest_ai::process_join(state, socket, data).await;
}

async fn handle_join_room(_state: &AppState, socket: &SocketRef, room: &str) {
    let room_owned = room.to_string();
    let _ = socket.join(room_owned.clone());
    ROOM_SESSIONS
        .entry(room_owned.clone())
        .or_default()
        .push(socket.id.to_string());
    let _ = socket.emit("user_joined", &json!({"room": room_owned}));
}

async fn handle_chat(state: &AppState, socket: &SocketRef, data: Value) {
    guest_ai::process_chat(state, socket, data).await;
}

async fn handle_chat_history(state: &AppState, socket: &SocketRef, room: &str) {
    let history = state.chat_service.get_history("default", room);
    let _ = socket.emit("chat_history", &Value::Object(history));
}

async fn handle_chat_history_guest(state: &AppState, socket: &SocketRef, data: Value) {
    let app_id = data
        .get("appId")
        .or_else(|| data.get("app_id"))
        .and_then(|v| v.as_str())
        .unwrap_or("default");
    let guest_session_id = data
        .get("guestSessionId")
        .or_else(|| data.get("guest_session_id"))
        .and_then(|v| v.as_str());
    let guest_phone = data
        .get("guestPhone")
        .or_else(|| data.get("guest_phone"))
        .and_then(|v| v.as_str());

    if guest_session_id.is_none() && guest_phone.is_none() {
        let _ = socket.emit("chat_history_guest", &json!([]));
        return;
    }

    let history = state.chat_service.get_history_by_guest_identity(
        app_id,
        guest_session_id,
        guest_phone,
        50,
    );
    let _ = socket.emit("chat_history_guest", &json!(history));
}

async fn handle_chat_history_app(state: &AppState, socket: &SocketRef, app_id: &str) {
    let history = state.chat_service.get_history(app_id, "app");
    let _ = socket.emit("chat_history_app_snapshot", &Value::Object(history));
}

async fn handle_csm_sign_in(_state: &AppState, socket: &SocketRef, _encrypted: &str) {
    let _ = socket.emit("csm_sign_in", &json!({"ok": true, "message": "sign-in via socket"}));
}

fn list_users(state: &AppState, app_id: &str) -> Value {
    let page = state.record_manager.filter(app_id, "csm_accounts", &Default::default());
    page.get("data").cloned().unwrap_or(json!([]))
}

File: backend-go/internal/socket/events.go
package socket

import (
	"encoding/json"
	"log"
	"strings"
	"sync"
	"time"

	socketio "github.com/zishang520/socket.io/v2/socket"

	"csm_server/backend-go/internal/model"
	"csm_server/backend-go/internal/services"
)

type sessionState struct {
	appID         string
	username      string
	guestPhone    string
	guestIdentity string
	locale        string
	isAdmin       bool
}

var (
	sessionMu sync.RWMutex
	sessions  = map[string]*sessionState{}
)

func registerSocketEvents(s *socketio.Socket, io *socketio.Server, deps Dependencies) {
	s.On("join_room", func(args ...any) {
		room := stringArg(args, 0)
		if room == "" {
			return
		}
		s.Join(socketio.Room(room))
		_ = s.Emit("user_joined", map[string]any{"room": room})
	})

	s.On("join", func(args ...any) {
		data := mapArg(args, 0)
		handleJoin(s, data)
	})

	s.On("chat", func(args ...any) {
		data := mapArg(args, 0)
		handleChat(s, deps, data)
	})

	s.On("csm_msg_update", func(args ...any) {
		if len(args) > 0 {
			s.Broadcast().Emit("csm_msg_update", args[0])
		}
	})

	s.On("chat_history", func(args ...any) {
		room := stringArg(args, 0)
		history := deps.Chat.GetHistory("default", room)
		_ = s.Emit("chat_history", history)
	})

	s.On("chat_history_guest", func(args ...any) {
		data := mapArg(args, 0)
		appID := firstStr(data, "appId", "app_id", "default")
		sid := firstStr(data, "guestSessionId", "guest_session_id")
		phone := firstStr(data, "guestPhone", "guest_phone")
		if sid == "" && phone == "" {
			_ = s.Emit("chat_history_guest", []any{})
			return
		}
		history := deps.Chat.GetHistoryByGuestIdentity(appID, sid, phone, 50)
		_ = s.Emit("chat_history_guest", history)
	})

	s.On("chat_history_app", func(args ...any) {
		appID := stringArg(args, 0)
		if appID == "" {
			appID = "default"
		}
		history := deps.Chat.GetHistory(appID, "app")
		_ = s.Emit("chat_history_app_snapshot", history)
	})

	s.On("request_chat_history_app_snapshot", func(args ...any) {
		appID := stringArg(args, 0)
		if appID == "" {
			appID = "default"
		}
		history := deps.Chat.GetHistory(appID, "app")
		_ = s.Emit("chat_history_app_snapshot", history)
	})

	s.On("chat_mark_read", func(args ...any) {
		_ = s.Emit("chat_mark_read", map[string]any{"ok": true})
	})

	s.On("chat_mark_all_read", func(_ ...any) {
		_ = s.Emit("chat_mark_all_read", map[string]any{"ok": true})
	})

	s.On("chat_recall_message", func(args ...any) {
		if len(args) > 0 {
			io.Emit("chat_message_recalled", args[0])
		}
	})

	s.On("user_typing", func(args ...any) {
		if len(args) > 0 {
			s.Broadcast().Emit("user_typing", args[0])
		}
	})

	s.On("broadcast_notification", func(args ...any) {
		if len(args) > 0 {
			io.Emit("notification", args[0])
		}
	})

	s.On("register_guest_phone", func(args ...any) {
		data := mapArg(args, 0)
		handleRegisterGuestPhone(s, deps, data)
	})

	s.On("chat_guests_list", func(args ...any) {
		appID := stringArg(args, 0)
		_ = s.Emit("chat_guests_list", deps.Chat.GetGuestSessionsByAppID(appID))
	})

	s.On("chat_list_users", func(args ...any) {
		appID := stringArg(args, 0)
		page := deps.RM.Filter(appID, "csm_accounts", model.SearchFilter{Operator: "AND"})
		_ = s.Emit("chat_list_users", page["data"])
	})

	s.On("csm_sign_in", func(_ ...any) {
		_ = s.Emit("csm_sign_in", map[string]any{"ok": true, "message": "sign-in via socket"})
	})

	s.On("csm_register_an_account", func(_ ...any) {
		_ = s.Emit("csm_register_an_account", map[string]any{"ok": true})
	})
}

func handleJoin(s *socketio.Socket, data map[string]any) {
	sid := string(s.Id())
	appID := firstStr(data, "appId", "app_id")
	if appID == "" {
		appID = "csm"
	}
	username := firstStr(data, "username")
	if username == "" {
		username = "Guest"
	}
	userID := firstStr(data, "userId", "user_id")
	isAdmin, _ := data["isAdmin"].(bool)
	guestPhone := firstStr(data, "guestPhone", "guest_phone")
	guestSessionID := firstStr(data, "guestSessionId", "guest_session_id")
	locale := firstStr(data, "locale")
	if locale == "" {
		locale = "vi"
	}

	st := &sessionState{
		appID: appID, username: username, guestPhone: guestPhone,
		guestIdentity: guestSessionID, locale: locale, isAdmin: isAdmin,
	}
	sessionMu.Lock()
	sessions[sid] = st
	sessionMu.Unlock()

	if isAdmin {
		masterRoom := "app:" + appID
		s.Join(socketio.Room(masterRoom))
		_ = s.Emit("user_joined", map[string]any{"room": masterRoom, "username": username})
		return
	}

	if guestSessionID != "" || guestPhone != "" {
		identity := guestSessionID
		if identity == "" {
			identity = guestPhone
		}
		if identity == "" {
			identity = sid
		}
		st.guestIdentity = identity
		privateRoom := "guest:" + appID + ";" + identity
		masterRoom := "app:" + appID
		s.Join(socketio.Room(privateRoom))
		s.Join(socketio.Room(masterRoom))
		_ = s.Emit("user_joined", map[string]any{"room": privateRoom, "username": username})
		return
	}

	room := firstStr(data, "room")
	if room == "" {
		room = appID
	}
	s.Join(socketio.Room(room))
	_ = s.Emit("user_joined", map[string]any{"room": room})

	if userID != "" {
		authRoom := "auth:user:" + userID
		s.Join(socketio.Room(authRoom))
		_ = s.Emit("user_joined", map[string]any{"room": authRoom, "username": username})
	}
}

func handleChat(s *socketio.Socket, deps Dependencies, data map[string]any) {
	if data == nil {
		return
	}
	sid := string(s.Id())
	sessionMu.RLock()
	st := sessions[sid]
	sessionMu.RUnlock()

	appID := firstStr(data, "appId", "app_id")
	if appID == "" && st != nil {
		appID = st.appID
	}
	if appID == "" {
		appID = "csm"
	}
	data["appId"] = appID

	isAdmin, _ := data["isAdmin"].(bool)
	if !isAdmin && st != nil && st.isAdmin {
		isAdmin = true
	}
	userID := firstStr(data, "userId", "user_id")
	room, _ := data["room"].(string)
	guestContext := strings.HasPrefix(room, "guest:") || userID == ""

	if isAdmin && guestContext {
		// admin replying in guest context — persist + broadcast
	} else if userID == "" {
		guestIdentity := firstStr(data, "guestSessionId", "guest_session_id")
		if guestIdentity == "" && st != nil {
			guestIdentity = st.guestIdentity
		}
		if guestIdentity == "" {
			guestIdentity = sid
		}
		data["guestSessionId"] = guestIdentity
		if _, ok := data["username"]; !ok {
			data["username"] = "Khach hang"
		}
		data["room"] = "guest:" + appID + ";" + guestIdentity
		if st != nil && st.guestPhone != "" && firstStr(data, "guestPhone", "guest_phone") == "" {
			data["guestPhone"] = st.guestPhone
		}
	}

	if data["timestamp"] == nil {
		data["timestamp"] = time.Now().UnixMilli()
	}

	if err := deps.Chat.SaveMessage(appID, data); err != nil {
		log.Printf("Socket chat save failed: %v", err)
	}
	s.Broadcast().Emit("message", data)

	if guestContext && !isAdmin {
		message := firstStr(data, "message", "text", "content")
		locale := "vi"
		sessionMu.RLock()
		if st != nil && st.locale != "" {
			locale = st.locale
		}
		sessionMu.RUnlock()
		if strings.TrimSpace(message) != "" {
			go emitGuestAiReply(s, deps, appID, data, message, locale)
		}
	}
}

func handleRegisterGuestPhone(s *socketio.Socket, deps Dependencies, data map[string]any) {
	appID := firstStr(data, "appId", "app_id")
	oldIdentity := firstStr(data, "guestSessionId", "guest_session_id")
	phone := firstStr(data, "phone")
	if appID == "" || oldIdentity == "" || phone == "" {
		_ = s.Emit("register_guest_phone", map[string]any{"success": false, "error": "Missing required fields"})
		return
	}
	rebound := deps.Chat.RebindGuestPhone(appID, oldIdentity, phone)
	permanentRoom := "guest:" + appID + ";" + phone
	s.Join(socketio.Room(permanentRoom))
	sessionMu.Lock()
	if st := sessions[string(s.Id())]; st != nil {
		st.guestIdentity = phone
		st.guestPhone = phone
	}
	sessionMu.Unlock()
	_ = s.Emit("register_guest_phone", map[string]any{
		"success": true, "phone": phone, "permanentRoom": permanentRoom, "rebound": rebound,
	})
}

func stringArg(args []any, idx int) string {
	if idx >= len(args) {
		return ""
	}
	switch v := args[idx].(type) {
	case string:
		return v
	default:
		return ""
	}
}

func mapArg(args []any, idx int) map[string]any {
	if idx >= len(args) {
		return nil
	}
	switch v := args[idx].(type) {
	case map[string]any:
		return v
	case string:
		var m map[string]any
		if json.Unmarshal([]byte(v), &m) == nil {
			return m
		}
	}
	return nil
}

func firstStr(m map[string]any, keys ...string) string {
	for _, k := range keys {
		if v, ok := m[k].(string); ok && strings.TrimSpace(v) != "" {
			return strings.TrimSpace(v)
		}
	}
	return ""
}

func emitGuestAiReply(s *socketio.Socket, deps Dependencies, appID string, guestMsg map[string]any, message, locale string) {
	reply := services.GuestChatReply(deps.Config, deps.Llama, message, locale, appID)
	if strings.TrimSpace(reply) == "" {
		return
	}
	room := firstStr(guestMsg, "room")
	if room == "" {
		room = "guest:" + appID
	}
	replyData := map[string]any{
		"room": room, "appId": appID, "message": reply,
		"username": "CSM AI", "isBot": true, "isAdmin": false,
		"timestamp":      time.Now().UnixMilli(),
		"guestSessionId": firstStr(guestMsg, "guestSessionId", "guest_session_id"),
	}
	if err := deps.Chat.SaveMessage(appID, replyData); err != nil {
		log.Printf("Guest AI reply save failed: %v", err)
	}
	s.Broadcast().Emit("message", replyData)
}
